usr/sbin/p11sak/p11sak.c

```diff
@@ -1143,29 +1137,32 @@ static CK_RV tok_key_list_init(CK_SESSION_HANDLE session, p11sak_kt kt,
     case kt_GENERIC:
     case kt_RSAPKCS:
     case kt_EC:
-        tmplt[1].type = CKA_KEY_TYPE;
-        tmplt[1].pValue = &a_key_type;
-        tmplt[1].ulValueLen = sizeof(CK_KEY_TYPE);
+        tmplt[count].type = CKA_KEY_TYPE;
+        tmplt[count].pValue = &a_key_type;
+        tmplt[count].ulValueLen = sizeof(CK_KEY_TYPE);
+        count++;
         break;
     case kt_SECRET:
     case kt_PUBLIC:
     case kt_PRIVATE:
-        tmplt[1].type = CKA_CLASS;
-        tmplt[1].pValue = &a_cko;
-        tmplt[1].ulValueLen = sizeof(CK_OBJECT_CLASS);
+        tmplt[count].type = CKA_CLASS;
+        tmplt[count].pValue = &a_cko;
+        tmplt[count].ulValueLen = sizeof(CK_OBJECT_CLASS);
+        count++;
+        break;
+    case kt_ALL:
         break;
     default:
         fprintf(stderr, "Unknown key type\n");
         return CKR_ARGUMENTS_BAD;
     }
 
     if (label != NULL_PTR) {
-        tmplt[2].type = CKA_LABEL;
-        tmplt[2].pValue = label;
-        tmplt[2].ulValueLen = strlen(label);
-        count = 3;
-    } else
-        count = 2;
+        tmplt[count].type = CKA_LABEL;
+        tmplt[count].pValue = label;
+        tmplt[count].ulValueLen = strlen(label);
```

Comment:
Listing keys by label may behave somehow strange, dependent how the label of the objects have been created.

With the current code, the label filter only matches when the object's label attribute matches exactly the specified label, i.e. the label attribute's ulValueLen is exactly the string length of the label.

In case the label of the object was created like "ABC   ", i.e. padded with blanks, then a label filter of "ABC" does not match and the objects is not listed. 

While this is technically correct, it might be confusing for users. 

However, I don't see an easy way how to solve this. We would need to filter manually by label, and trim all blanks from then end before comparing the strings.....
Not sure if this is the way to go...... Probably not for this PR.

Another option would be to quote the label when displayed by the list command (e.g. like 'ABC   ', so that at least the padded blanks can be seen in the list output. Then one can specify a label filter of `--label "ABC   "` which will then match the object's label.

Reply:
https://docs.microsoft.com/de-de/dotnet/api/system.string.trim?view=net-6.0#system-string-trim

Maybe something like this at key gen? Should be in another commit though

Reply:
well, the thing is that you don't know who creates the object. If it is created by p11sak, then its already OK (I guess), but if the object is created by some other application, you don't know (and have no influence) on how it is done.

I would vote for simply quoting the label value at the list command (for both short and long format), so that the training blanks at least can be seen.